docs/internals/data_file.md

```diff
@@ -0,0 +1,196 @@
+---
+sidebar_position: 2
+---
+
+# Data File
+
+> “Just show me the tables already!”
+> — probably not Fred Brooks
+
+Each TigerBeetle replica stores all data inside a single file, called the data file (conventional
+extension is `.tigerbeetle`). This document describes the high level layout of the data file. The
+presentation is simplified a bit, to provide intuition without drowning the reader in details.
+Consult the source code for byte-level details!
+
+The data file is divided into several zones, with the main ones being:
+
+- write-ahead log
+- superblock
+- grid
+
+The grid forms the bulk of the data file (up to several terabytes). It is an array of 64KiB blocks:
+
+```zig
+pub const Block = [constants.block_size]u8;
+pub const BlockPtr = *align(constants.sector_size) Block;
+```
+
+The grid serves as a raw storage layer. Higher level data structures (notably, the LSM tree) are
+mapped to physical grid blocks. Because TigerBeetle is deterministic, the contents of the grid is
+identical across up-to-date replicas. This storage determinism is exploited to implement state sync
+and repair on the level of grid blocks, see [the repair protocol](./vsr.md#protocol-repair-grid).
```

Comment:
```suggestion
and repair on the level of grid blocks, see [the repair protocol](./vsr.md#protocol-repair-grid).
TigerBeetle guarantees equality only for logical, used blocks, not for padding in used blocks
or free blocks.
```

Reply:
Hm, I think we guarantee that the padding in used blocks is zero? We always write the whole grid blocks, and we explicitly zero the padding.